src/vset.c

```diff
@@ -0,0 +1,2094 @@
+ *       • The input vector has only 1 or fewer elements (nothing to split).
+ *       • The `split_index` is equal to the vector length (all elements stay in the left part).
+ *       • The `split_index` is such that the right part would have 0 elements.
+ *
+ * Side effects:
+ *   - The original vector pointer (`*pv_ptr`) is modified to point to the
+ *     resized left portion.
+ *
+ * Example:
+ * --------
+ * Suppose `pv_ptr` points to a vector of 5 elements:
+ *     [A, B, C, D, E]
+ *
+ * Calling:
+ *     pVector *right = pvSplit(&pv_ptr, 3);
+ *
+ * Results in:
+ *     pv_ptr -> [A, B, C]
+ *     right   -> [D, E]
+ *
+ * If the split_index is 5 (i.e. the end), the function returns NULL and the
+ * original vector is unchanged. */
+pVector *pvSplit(pVector **pv_ptr, uint32_t split_index) {
+    pVector *pv = *pv_ptr;
+
+    // Handle edge cases: null or empty
+    if (!pv || pv->len <= 1) return NULL;
+
+    // If no valid split found, return NULL (entire vector is one block)
+    if (split_index == pv->len) return NULL;
+
+    // Number of elements for the right half
+    uint64_t right_len = pv->len - split_index;
+    if (right_len == 0) return NULL;
+
+    // Allocate new vector for right part
+    size_t item_bytes = sizeof(void *);
+    size_t total_bytes = sizeof(pVector) + right_len * item_bytes;
+    size_t new_alloc;
+    pVector *right = zmalloc_usable(total_bytes, &new_alloc);
+    right->alloc = new_alloc;
+    right->len = right_len;
+
+    // Copy the right part
+    memcpy(&right->data[0], &pv->data[split_index], right_len * item_bytes);
+
+    // Shrink original vector
+    pv->len = split_index;
+    *pv_ptr = pvShrinkToFit(pv); // Optional: shrink in-place to reduce memory
+
+    return right;
+}
+
+/* Creates a new pVector with the specified initial capacity.
+ *
+ * This function initializes a new pVector capable of holding at least
+ * `capacity` elements. Internally, it delegates allocation and setup to
+ * `pvMakeRoomFor`, starting from a NULL vector.
+ *
+ * Arguments:
+ *   capacity - The initial number of elements the vector should be able to store.
+ *
+ * Return:
+ *   A pointer to the newly allocated pVector.
+ *   Note that a NULL is a !!valid!! cector which size is zero.
+ *
+ * Note:
+ *   The logical length (`len`) of the returned vector is initialized to 0.
+ */
+pVector *pvNew(uint32_t capacity) {
+    pVector *new_vec = NULL;
+    return pvMakeRoomFor(new_vec, capacity);
+}
+
+/* Inserts an element at the specified position in the pVector.
+ *
+ * Ensures enough capacity for the new element, shifts elements to make space,
+ * and inserts the given element at the desired position.
+ *
+ * Arguments:
+ *   pv   - The pVector to insert into (can be NULL).
+ *   elem - The pointer to be inserted.
+ *   pos  - The index at which to insert the element (must be ≤ pv->len).
+ *
+ * Return:
+ *   The updated pVector with the element inserted. */
+pVector *pvInsert(pVector *pv, void *elem, uint32_t pos) {
```

Comment:
Suggest assertion to validate `pos`.  Otherwise it's possible for line 242 to index off the end.
```suggestion
pVector *pvInsert(pVector *pv, void *elem, uint32_t pos) {
    assert(pos <= pv->len);
```